Added time based sleep timers as well as episode count sleep timers

eblis opened the pull request:
### Description
Reimplementation of PR #7247, but with changes from PR #7713.
The sleep timers no longer have a thread, they rely on `PlaybackPositionEvent` event to be notified about current playback.

This pull request adds on top of the features made in #7713 and also adds Episode sleep timer.
I kept the "new" (added in #7247)  `PlaybackEndedEvent` event as a reliable way to determine when an episode completed. I played around a bit with the duration and remaining time in `PlaybackPositionEvent` but it seemed unreliable.

This has a few minor changes compared to the old pull request #7247, notably:
- When using episode sleep timers, playback no longer finishes on the next episode, it stops at the end of the last episode (like time based sleep timers do)
- Similar to time based sleep timers, the last few seconds of the last episode will now reduce volume right before stopping.
- There is now a difference between shown timers (display timers, which can be episode counts or seconds) and remaining duration (which is always milis) - in the previous pull request we were using milis everywhere and the episode counter class was trying to be smart about the values and convert from milis to episodes - now we provide 2 values, one to show to the user and another one to use as milis duration.

### Checklist
<!-- 
  To help us keep the issue tracker clean and work as efficient as possible,
  please make sure that you have done all of the following.
  You can tick the boxes below by placing an x inside the brackets like this: [x]
-->
- [x] I have read the contribution guidelines: https://github.com/AntennaPod/AntennaPod/blob/develop/CONTRIBUTING.md#submit-a-pull-request
- [x] I have performed a self-review of my code
- [x] I have run the automated code checks using `./gradlew checkstyle spotbugsPlayDebug spotbugsDebug :app:lintPlayDebug`
- [x] My code follows the style guidelines of the AntennaPod project: https://github.com/AntennaPod/AntennaPod/wiki/Code-style
- [x] I have mentioned the corresponding issue and the relevant keyword (e.g., "Closes: #xy") in the description (see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] If it is a core feature, I have added automated tests

eblis commented:
I have updated this merge request after the merge of #7713 into develop.

This merge request adds episode timers as another option (the other being the already present clock sleep timer).

Adding a new sleep timer
<img width="384" height="473" alt="image" src="https://github.com/user-attachments/assets/90e208d8-958c-41e5-93aa-c42696ae891d" />

Showing configured episode sleep timer
<img width="376" height="502" alt="image" src="https://github.com/user-attachments/assets/ec4645d1-e9b0-4607-9b27-c589923444ce" />

Showing that there's a sleep timer active, if it's ending this episode and the amount of time left. I can put this into a configurable option if it's easier for this merge to be accepted if the option is configurable, just let me know.
<img width="422" height="78" alt="image" src="https://github.com/user-attachments/assets/a5c6873a-ca06-4edc-87be-2e6ff702a470" />

If it's showing remaining time and in orange then a sleep timer is active. If there's a clock icon too that means that the sleep timer is going to end (sometime) this episode - works for both clock and episode sleep timers.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -40,8 +51,65 @@ public class SleepTimerDialog extends DialogFragment {
+    private Spinner sleepTimerType;
     private CheckBox chAutoEnable;
 
+    Button extendSleepFiveMinutesButton;
+    Button extendSleepTenMinutesButton;
+    Button extendSleepTwentyMinutesButton;
```
This seems like a good opportunity to migrate the dialog to ViewBindings. See for example `QuickFeedDiscoveryFragment`

eblis replied:
I'd keep this for a later PR if that's ok with you, in case I mess up something and it will just add to the changes in this PR

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -40,8 +51,65 @@ public class SleepTimerDialog extends DialogFragment {
+        return map;
+    }
+
+    LinkedHashMap<SleepTimerType, SleepTimeConfig> allSleepConfig = linkedMapOf(List.of(
```
This feels more generic than it needs to be. I don't think there will be additional types of sleep timers anytime soon. Making the code less generic and more "manual" would make it easier to understand for future developers.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -169,6 +272,34 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+            try {
+                button.setText(getString(selectedConfig.buttonTextResourceId, entryConfig.displayValue));
+            } catch (Resources.NotFoundException ex) {
+                button.setText(getContext().getResources()
```
Exceptions as normal control flow are a bit unusual. I think the easier option would be to just set the text explicitly

eblis replied:
Not sure what you mean here, I've extracted the `setText` functionality to a separate method.

eblis replied:
I wanted to change the string for the minutes to a plural string as well, and even made the required changes, but reverted because all of the translations for this string will become invalid (for little to no gain).

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/audio/AudioPlayerFragment.java:
```diff
@@ -370,17 +374,44 @@ public void updatePosition(PlaybackPositionEvent event) {
+                final CharSequence text = txtvLength.getText();
+                txtvLength.setText(text + "");
+            }
+            txtvLength.setTextColor(
+                    MaterialColors.getColor(getContext(),
+                            android.R.attr.colorActivatedHighlight, Color.RED));
```
Please move this to a separate PR. We are currently redesigning the player screen, so it is not yet clear whether we want to add this or not.

ByteHamster commented on event/src/main/java/de/danoeh/antennapod/event/playback/SleepTimerUpdatedEvent.java:
```diff
@@ -2,37 +2,43 @@
 public class SleepTimerUpdatedEvent {
     private static final long CANCELLED = Long.MAX_VALUE;
-    private final long timeLeft;
+    private final long milisTimeLeft;
+    private final long displayTimeLeft;
```
What is the difference between displayTime and milisTime?

eblis replied:
There's no difference really for sleep timers based on seconds, but there's a difference for episode count sleep timers.
The sleep timer logic dealt with milliseconds until expiry, i left that in place, and the episode sleep timer uses a big number for the remaining time, 1 day in millis for each episode left - unless it's the last episode and we're told to stop after it, then it uses the remaining time in the current episode.

The sleep timer logic uses these milliseconds to see how long it has left and whether it should reduce the volume for the last few seconds - works the same for both episodes and time based sleep timers.

However, when the sleep timer dialog is open the code wanted to convert the millisecond timer into pretty time for the user, but that doesn't really work well with the episode counter. So instead of passing just milliseconds around I created a new class that can have both a pretty value and the millisecond value used for all computations. The pretty value is used when showing the dialog to the user, for episode timers it knows to show 1, 2, 3, .. episodes, for timer based sleep timers it formats the millseconds to a pretty time format.

(Sidenote: the original PR for this didn't have the 2 values, instead it relied on the fact that episode timers always reported days in millis and converted back, but that felt more of a hack and I added the new class instead in the new PR)

ByteHamster commented on playback/service/src/main/java/de/danoeh/antennapod/playback/service/internal/EpisodeSleepTimer.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    @SuppressWarnings("unused")
+    public void onPlaybackEnded(PlaybackEndedEvent playbackEndedEvent) {
+        handlePlayedEpisode();
+    }
```
Doing this with events could cause race conditions. Please "ask" the sleep timer explicitly when AntennaPod decides whether to play the next episode (evaluating the "continuous playback" setting, queue availability, etc)

eblis replied:
Changed the code to not use the `PlaybackEndedEvent` anymore (also removed the event).
Instead the code will now directly notify the sleep timer (if available) when an episode finished playback.
It will also ask the sleep timer if playback can continue or not.

ByteHamster replied:
Sounds good. Could you push the change please? :)

ByteHamster commented on playback/service/src/main/java/de/danoeh/antennapod/playback/service/internal/TimerValue.java:
```diff
@@ -0,0 +1,19 @@
+package de.danoeh.antennapod.playback.service.internal;
+
+public class TimerValue {
+    private final long displayValue;
+    private final long milisValue;
```
Wouldn't this be easier to understand if it had "type" and "value" instead of display/milis? I assume encoding the type in the value is the reason why you need two different ones?

eblis replied:
For time based sleep timers `displayValue` and `millisValue` are the same value, say 245245 millis left, but for episode sleep timers they're different, `displayValue` may be 1 (episode left), whereas `millisValue` will be the milliseconds left in the current episode.

ByteHamster replied:
Okay, maybe a comment after displayValue could help.

```suggestion
    private final long displayValue; // Value shown to user (milliseconds or number of episodes)
    private final long milisValue;
```

keunes commented:
Thanks for your work on this! We discussed it in the Needs Decision meeting.

There's a bit of a risk of confusion if 'continuous playback' is _enabled_ and sleep timer's 'end after x episodes' is active (see https://github.com/AntennaPod/AntennaPod/issues/2146#issuecomment-1272016332). We had two options:
* Not implement any conditions/restrictions: in case of questions or comments like 'it's not working' we'd have to ask the user what they've set in both places which is annoying
* If we implement a condition: in case of questions or comments like 'why can't I do x' we have always the same clear answer "because there's this other setting for that" (which is easier and we could even have a simple documentation article if it gets asked a lot). The condition/changes would be:
  * Make sleep timer 'auto enable' setting always clickable (currently it can only be tapped if 'auto enable' is active)
  * When the auto enable timing is set to 'Always' then the sleep timer 'auto enable' setting is inactive and disabled. ~~In the background, the original value (on/off) is kept so that if you change the sleep timer auto-enable timeslot the original value is applied again.~~ (This would be nice for users but given that it's  a niche use-case we can just overwrite the setting to 'off' when the timeslot 'Always' is applied.)
  * When the 'auto enable' setting is deactivated and the user taps on it, a snackbar should be shown with a brief explanation: "Change the 'continous playback' setting if you always want to stop playback after an episode"

Would you be OK implementing that as well @eblis?
  
And we need to prevent confusion if continuous playback is _disabled_:
* disallow sleep timer episode > 1, either by:
  * showing warning if number > 1 is entered, or
  * disabling & fixing number field to 1 (preferred)
* use radio buttons (but busy UX and not possible to set sleep timer to > 1 episode at all)
* if sleep timer is > 1 episode still continue to run the set episodes despite disabled continuous playback

What do you think @eblis? Personally I would go for the last option but @ByteHamster said that's creating a bit of a mess in the code.

Matth7878 commented:
> * if sleep timer is > 1 episode still continue to run the set episodes despite disabled continuous playback

IMHO it's OK to do this even when continuous playback is disabled. In these case there is an explicit action from the user to stop only after a number of episodes.

eblis commented:
> Would you be OK implementing that as well @eblis?

I am currently on holiday, but will look at implementing all the suggestions once I'm back, probably at the start of September.

eblis commented:
I have made some changes to this pull request.
* When an episode sleep timer is active playback will continue for the number of episodes left in the sleep timer. This will happen even if _continuous playback_ is disabled, as the action to sleep after X episodes was selected by the user so the idea was it overrides the continuous playback flag.
* Sleep timer dialog will now let you change the _to_ and _from_ automatic sleep timer enablement without first checking the flag before it - if the user selects a time then the flag to "auto enable" will be automatically selected as well.
  * But if the user selects "Always" as the interval then the "auto enable" checkbox will get disabled.
  * However, if the user clicks on the "auto enable" flag when the interval is set to "Always" a snackbar appears and tells them "Change the 'continous playback' setting if you always want to stop playback after an episode" (only appears when they enable the button and the interval is set to Always - let me know if you want to show this message more often), but will let them enable the flag.
*  Reworked the sleep timer dialog types so they're less generic, but easier to follow. Only CLOCK and EPISODES types for now.
* Removed the `PlaybackEndedEvent` and instead reworked the code so whenever playback finishes for an event the sleep timer is immediately notified that an episode has finished playback and then the sleep timer is asked if playback can continue (this is used to allow the playback to continue for the configured number of episodes even if _continuous playback_ flag is disabled).

ByteHamster commented on event/src/main/java/de/danoeh/antennapod/event/playback/SleepTimerUpdatedEvent.java:
```diff
@@ -2,37 +2,43 @@
-    public long getTimeLeft() {
-        return Math.abs(timeLeft);
+    public long getMilisTimeLeft() {
+        return Math.abs(milisTimeLeft);
```
There is the second `l` missing in `millis`

ByteHamster commented:
Thanks for updating the PR, and sorry for taking a month to reply. This PR is more complex than others, so I was a bit lazy and first looked at simpler things.

> if sleep timer is > 1 episode still continue to run the set episodes despite disabled continuous playback

Hmm, I'm not super convinced by that, to be honest. I would just show an error message saying that continuous playback is disabled.

> Are we trying to prevent the user from setting the whole day as a sleep timer interval and guiding them to use the "Continuous playback" setting, if that is what they are after?

The idea is to prevent users using the sleep timer with 1 episode and always automatically activated instead of just turning off "continuous playback"

> However, if the user clicks on the "auto enable" flag when the interval is set to "Always" a snackbar appears and tells them "Change the 'continous playback' setting if you always want to stop playback after an episode" (only appears when they enable the button and the interval is set to Always - let me know if you want to show this message more often), but will let them enable the flag.

Maybe also when saving the dialog? I will need to play around with that.

eblis commented on playback/service/src/main/java/de/danoeh/antennapod/playback/service/PlaybackService.java:
```diff
@@ -1512,11 +1549,15 @@ public boolean sleepTimerActive() {
     }
 
-    public long getSleepTimerTimeLeft() {
+    public boolean isSleepTimerEndingThisEpisode(long episodeRemainingMillis) {
```
It's used to change the color of the time left in the UI, but I removed that part (to add as separate PR after this one)

eblis commented:
> The idea is to prevent users using the sleep timer with 1 episode and always automatically activated instead of just turning off "continuous playback"

This is actually what I do 😄, but only during the night. 
So if I start a podcast during the configured time interval, it will automatically enable sleep timer for 1 episode.
If I start during the day it will play as many as I'll let it.

> Hmm, I'm not super convinced by that, to be honest. I would just show an error message saying that continuous playback is disabled.

It's based on @Matth7878's suggestion, as the user explicitly said to play x episodes, so we should honor that.

keunes commented:
Thanks for your continued work on this @eblis !

>>> if sleep timer is > 1 episode still continue to run the set episodes despite disabled continuous playback
>>
>>Hmm, I'm not super convinced by that, to be honest. I would just show an error message saying that continuous playback is disabled.
>
> It's based on @Matth7878's suggestion, as the user explicitly said to play x episodes, so we should honor that.

The user also actively decided (via the settings) _not_ to have continued playback (it is enabled by default so if it's disabled they made a conscious choice). I think we should not allow the user to set contradicting settings.  It'll probably cause (support) questions later on.

I agree with @ByteHamster therefore that we should just show a warning.

Given that continuous playback is enabled by default, I suspect that few users will see the warning.

Matth7878 commented:
> I think we should not allow the user to set contradicting settings. It'll probably cause (support) questions later on.

Yet you voted up on my reply! 😝

I don't agree with you : either it should not be possible in first place, either you allow it as it's an immediate action from user telling that right now he wants to listen up to x episodes even though he disabled continuous playback. Why telling him he can't when he just asked for it and confirmed it via starting sleep timer?

Would someone really complains because he choose to listen to 2 episodes and then say why would AntennaPod let me do so? I believe it's more likely to have complain about being forced to go through the pain to disable an option in settings just to be allowed once to listen to more than one episode. And afterward to remember to change back setting.

eblis commented:
> True, but IMHO that's OK because 'stopping earlier' (stop automatically, before the episode/queue is finished) is the whole point of the sleep timer.

You could argue that the point of the sleep timer is to stop at a different time than normal, be it earlier if you have continuous playback enabled and you decided to stop after _x_ episodes, or be it later if you have continuous playback disabled, but want to play _y_ episodes at this point.

eblis commented:
I added a snackbar to alert the user
<img width="369" height="150" alt="image" src="https://github.com/user-attachments/assets/9c1ce038-af82-4b64-beec-282f5f179a05" />

keunes commented:
> You could argue that the point of the sleep timer is to stop at a different time than normal, be it earlier if you have continuous playback enabled and you decided to stop after x episodes, or be it later if you have continuous playback disabled, but want to play y episodes at this point.

That's a fair argument.

> if people start shooting themselves in the foot (asking why it didn't stop at the end of the episode when they selected a sleep timer of 2+ episodes, but have also disabled continuous playback etc) we change the code to not allow them to do that anymore - since it's a relatively simple change.

I'm kinda ok with that, but even if we take this approach, when we change the code to no longer allow this, we'll get complaints from users that we 'rempve functionality'. While if we just don't allow it as per @Matth7878 's suggestion, we don't take anything 'away' so folks are less likely to complain. At most, users will request to allow this, which we can also use as a barometer to see how much people need this…

So I don't know. WDYT @ByteHamster ?

ByteHamster commented:
I would also prefer to not allow it in the first place because users will complain when removing it later

eblis commented:
How should it work? 

Do we show a snack bar that they don't have continuous playback enabled so they can set it to 1 episode at most?

KraXen72 commented:
Can this accomodate the use-case where the 1-episode sleep timer will auto-enable between the set times?

I think I saw someone else mention it here as well, but basically, the idea for me is:
during the day, i want continuous playback to be on, and then when i go to sleep, i wanna put my phone on the charger (out of reach) and finish listening to the current episode
- if i fall asleep before the episode ends, which is unlikely, i know it won't keep playing through the night and can easily go through that 1 ep to find where I left off
- if i'm still awake when the episode ends, is auto-stops, signalling to me that it's time to go to sleep - it doesen't start the next one and get me interested in it

eblis commented:
> Actually, if "continuous playback" is disabled, do any of the "episodes" sleep timer values make any sense? Does starting the timer make any sense?

I don't think the episode timers make sense, even if you enable the timer or you don't, playback is going to finish at the end of the episode.


> How about disabling the entire screen (except for the dropdown/spinner and the error message you already added) when "continuous playback" is disabled?

Sounds good to me. 
Then the user can't configure episode sleep timers at all in this case, so the whole issues with the snackbar showing, allowing the user to continue are prevented as well, no?

eblis commented:
When continuous playback is off you get this view. I decided to keep the "set timer" button in, but disable it so the user has more hints that something is wrong, they shouldn't be doing this and are not allowed to. I thought that just hiding the button may make users start asking where the button is, why is it not visible, etc.

<img width="386" height="368" alt="image" src="https://github.com/user-attachments/assets/6aab4721-cdd9-4cff-bdd4-180ff9e0685b" />

Checking the "automatically enable timer ..." button will show just the snackbar, like it did before. I left it in since we talked about it a while ago.
<img width="386" height="465" alt="image" src="https://github.com/user-attachments/assets/5932ae6c-399e-4260-803f-0340260e52eb" />

But once you configure the time interval and select Always you get a new dialog asking you what you want to do.
<img width="404" height="344" alt="image" src="https://github.com/user-attachments/assets/18970201-da4b-4afc-82fd-f1224a4e955f" />
If you click on "Use sleep timer" it will let you set the episode sleep timer.
If you click on the "Disable continuous playback" button however it will disable continuous playback for you and then you get the first image, where you can't set an episode sleep timer.

The snackbar and the dialog are trying to do a similar thing and I didn't know the best way to leave both of them in.
Also, I didn't want to be intrusive and bother the users with the dialog if they're configuring "normal" intervals (i.e. not always), so that's why the dialog is showing only when they configure 24h interval.

EDIT: Maybe instead of the snackbar we could show the dialog instead in that case too, so the user is reminded that they have 24h interval and they can disable continuous playback .. the snackbar is only shown if 24h interval is set too (but currently for both timer and episode sleep timers)

eblis commented:
New dialog window. Also shown instead of the snackbar when the user checks "automatically activate the sleep timer ..." button.

<img width="386" height="345" alt="image" src="https://github.com/user-attachments/assets/095390de-8744-4be8-a5f4-4fb9695c67c4" />

ByteHamster commented on ui/i18n/src/main/res/values/strings.xml:
```diff
@@ -640,9 +644,24 @@
     <string name="sleep_timer_always">Always</string>
     <string name="sleep_timer_label">Sleep timer</string>
     <string name="time_dialog_invalid_input">Invalid input, time has to be an integer</string>
+    <string name="timer_exceeds_remaining_time_while_continuous_playback_disabled">Continuous playback is disabled and the current episode has only %1$d minutes left</string>
```
This should be plural. Some languages have different forms for 11, 21, 31, 41 minutes (not just a special case for 1 like in English)

ByteHamster commented on ui/i18n/src/main/res/values/strings.xml:
```diff
@@ -640,9 +644,24 @@
+        <item quantity="other">%d episodes</item>
+    </plurals>
+    <string name="sleep_timer_without_continuous_playback_title">Disable continuous playback</string>
+    <string name="sleep_timer_without_continuous_playback_message">You can simply disable continuous playback to automatically stop after an episode, there is no need to use the sleep timer for that.\n\nIf you actually want that to happen only at specific times you can set up a sleep timer.</string>
```
```suggestion
    <string name="sleep_timer_without_continuous_playback_message">By disabling \"continuous playback\" in the settings, you can automatically stop after each episode without using the sleep timer.</string>
```

ByteHamster commented on playback/service/src/main/java/de/danoeh/antennapod/playback/service/internal/EpisodeSleepTimer.java:
```diff
@@ -0,0 +1,74 @@
+
+        final TimerValue current = getTimeLeft();
+
+        if (isEndingThisEpisode(playbackPositionEvent.getPosition())) {
```
The function accepts remaining millis, not the position. However, doing this with the position/duration is difficult because of "skip silence", playback speed, and "skip ending". I would just ignore the volume reduction for episode sleep timers

ByteHamster commented on storage/preferences/src/main/java/de/danoeh/antennapod/storage/preferences/SleepTimerPreferences.java:
```diff
@@ -45,11 +49,26 @@ public static String lastTimerValue() {
+    }
+
+    public static SleepTimerType getSleepTimerType() {
+        return SleepTimerType.valueOf(prefs.getString(PREF_TIMER_TYPE, DEFAULT_TIMER_TYPE).toUpperCase(Locale.US));
```
Please add an "id" parameter to the enum values and store the integer instead. See for example `AutoDeleteAction`

ByteHamster commented on storage/preferences/src/main/java/de/danoeh/antennapod/storage/preferences/SleepTimerType.java:
```diff
@@ -0,0 +1,6 @@
+package de.danoeh.antennapod.storage.preferences;
+
+public enum SleepTimerType {
```
Move to `SleepTimer.Type`? This file is so simple, opening it is more overhead than just having it directly in the SleepTimer interface

eblis replied:
I tried doing that, but they're defined in different modules, we would need to add module dependencies. Is that ok?
Seems easier to just have the enum defined as it currently is.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -1,49 +1,129 @@
+        }
+    }
+
+    static class SleepTimeConfig {
```
I don't think we need this abstraction. Just do an if/else please.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -169,11 +288,156 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
     }
 
+    private void confirmAlwaysSleepTimerDialog() {
+        @SuppressLint("VisibleForTests")
```
Why is this needed? This is not a test

eblis replied:
The setter is marked as public for testing, this is the only other place that uses it outside of the tests.
IDE complains without the annotation and I didn't know if I could remove the annotation from the originating class.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -169,11 +288,156 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+        dialog.show();
+        // mark the disable continuous playback option in red
+        dialog.getButton(androidx.appcompat.app.AlertDialog.BUTTON_NEGATIVE)
+                .setTextColor(ContextCompat.getColor(requireContext(), android.R.color.holo_red_dark));
```
We have our own error color somewhere if I remember correctly. I think that color even adapts to the theme. I think `ThemeUtils.getColorFromAttr` or something like that should help

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -169,11 +288,156 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+
+    private void setButtonTextWithPlurals(Button button, int buttonTextResourceId, int value) {
+        button.setText("+" + getResources().getQuantityString(buttonTextResourceId, value, value));
+    }
```
There is just one caller of this function, just inline it.

eblis commented on playback/service/src/main/java/de/danoeh/antennapod/playback/service/internal/LocalPSMP.java:
```diff
@@ -678,9 +678,11 @@ public void onAudioFocusChange(final int focusChange) {
         releaseWifiLockIfNecessary();
 
+        callback.episodeFinishedPlayback(); // notify that the current episode just finished
+
```
To find out when an episode has finished playing.
What would be the alternative?

eblis commented on playback/service/src/main/java/de/danoeh/antennapod/playback/service/PlaybackService.java:
```diff
@@ -1076,7 +1090,11 @@ private Playable getNextInQueue(final Playable currentMedia) {
-        if (!UserPreferences.isFollowQueue()) {
+        // contiue playback if user has enabled continuous playback
+        // OR they enabled an episode sleep timer and there are still episodes left to play
+        final boolean continuousPlayback = UserPreferences.isFollowQueue() || shouldContinueToNextEpisode();
```
The callback is still needed though.
I can change the code to reference the sleep timer directly, but since the sleep timer might be `null` the condition will look weird I think, at least it's not easier to understand for me. 

Not have thought it through, but at first glance it would be something like this, provided we want to keep it on a single line:

```
final boolean continuousPlayback = UserPreferences.isFollowQueue()
   && (sleepTimer == null || !sleepTimer.shouldContinueToNextEpisode());
```

eblis replied:
I ended up with 
```
        final boolean continuousPlayback = UserPreferences.isFollowQueue()
                || (sleepTimer != null && sleepTimer.shouldContinueToNextEpisode());
```
Keep like this or revert to `final boolean continuousPlayback = UserPreferences.isFollowQueue() || shouldContinueToNextEpisode();` ?

ByteHamster replied:
I just pushed a few commits, including reverting back to call the function :)

eblis commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/SleepTimerDialog.java:
```diff
@@ -169,11 +237,149 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+            if (SleepTimerPreferences.getSleepTimerType() == SleepTimerType.EPISODES) {
+                // for episode timers check if the queue length exceeds the number of sleep episodes we have
+                final int queueSize = DBReader.getRemainingQueueSize(
+                        PlaybackPreferences.getCurrentlyPlayingFeedMediaId());
```
We need this to warn the user, per [this comment](https://github.com/AntennaPod/AntennaPod/pull/7841#issuecomment-3368109011)

eblis replied:
I'll try to get it to load in a separate thread. If the value loaded in time you'll get the warning, if the value is still default (`null`) you won't see the warning. I think it should work for most people, but then again I don't have a huge queue so I don't know how long it takes (but we're only grabbing the IDs, we don't care about all the queue item fields)

ByteHamster replied:
For threading please use rxjava instead of creating a thread manually. See other screens like subscriptions for an example

ByteHamster commented:
> Would it be possible to store/remember two separate values, and make the one for episodes one or two by default?

This could also be quite annoying if you type the number first and then change the unit, it would then change the number again

> Maybe the 'set sleep timer' button should be below the option

Not sure about this one. The idea is that you read it from top to bottom and usually don't need to look at the settings before enabling

> We disable automatic enabling when switching to 'episodes'

The idea is that if the setting is already enabled and you switch to "episodes", you miss the message about continuous playback

> I agree with this: "Instead of +1, +3, +5, maybe 1,2,3 would be better?" At least for when the sleep timer is set to 'Episodes'. There was a comment that it was implemented, but with episode sleep timer enabled it says "+n" episode(s).

This was about 123 vs 135 not about removing the "+"

> I feel that 'shake to reset' doesn't really make sense with the 'episodes' setting

👍 

> Actually, that makes me think; maybe it makes sense to show it (already) when a period of 13 hours or more is shown (i.e. the majority of the day).

Currently we do not let users select "always" at all for one episode. This is to avoid them ignoring the message and still setting up the timer. When showing the message even for 13 hours that would have to be different. I think I would just keep it as it is

eblis commented:
When switching between episodes and minutes I would just chance the value to the default, say 15 minutes and 1 episode timer.
We still only save one value, but when you switch type we revert back to the default for that type. 

I've been using episode sleep timers for about 10 months now and I've never switched type once

ByteHamster commented:
Right, I think optimizing for regularly switching units is additional code complexity that almost nobody would profit from. This PR is already long enough. I would leave that out and see if someone complains

keunes commented:
>> Would it be possible to store/remember two separate values, and make the one for episodes one or two by default?
>
> This could also be quite annoying if you type the number first and then change the unit, it would then change the number again

Not very likely to happen I think, as it's a new feature: you don't type 2 with in mind already the goal to set it to '2 episodes' – unless you already know that this functionality of 'episodes' exists from the release notes you've happened to read.

In future or subsequent cases, you know that it keeps its value.

That said, I would suspect many users pick one unit (ep or min) and stick with it, rather than frequently switch between them, in which case it's less of an issue.

>> Maybe the 'set sleep timer' button should be below the option
>
> Not sure about this one. The idea is that you read it from top to bottom and usually don't need to look at the settings before enabling

Ok, I'll create a separate issue so we can discuss further.

>> We disable automatic enabling when switching to 'episodes'
>
> The idea is that if the setting is already enabled and you switch to "episodes", you miss the message about continuous playback

But the message is only relevant if you have the automatic enabling set to 'always'. I would expect most users to actually have set a time slot. For these users it's now more annoying while they're not affected (they don't need to see the message).

Just like the message is shown immediately when saving the 'always' timeslot, we could add a condition to show it when the timeslot is already 'always' and the user switches to 'episodes'.
That way we warn the users who should be warned without hindering users that don't need to be.

>> Actually, that makes me think; maybe it makes sense to show it (already) when a period of 13 hours or more is shown (i.e. the majority of the day).
>
> Currently we do not let users select "always" at all for one episode. This is to avoid them ignoring the message and still setting up the timer. When showing the message even for 13 hours that would have to be different. I think I would just keep it as it is

It is possible actually; just press the back button of the OS.
But that's a separate thing; I'm proposing to in a way be more strict: show the warning/prevent this combination of settings not just when it's 'always' but also when the user tries to circumvent this by setting the timeslot from 05:00 to 04:00. By taking the criterion 'if set to Always or selected timeframe is 13 hours or more'.

keunes commented:
> I wanted to implement this, but couldn't decide on the cutoff time (i was hovering around 20 hours though) so I left it out.

My thinking was: sleep timer activation period makes most sense in the evening – from six to midnight if your going-to-bed-time varies a lot, that's 6 hours. Another scenario might be that you set this range not the time you go to bed, but that you include the time you wake up – for example from eight at night to eight in the morning. That's 12 hours.
Anything beyond that is closer to 'always' than a specific slot in the day, hence my proposal to show it when it's 13 hours or more.

We could also be more conservative of course (increasing the threshold value), but then can't think of a logical reasoning to determine the amount of hours.

> When switching between episodes and minutes I would just chance the value to the default, say 15 minutes and 1 episode timer.
We still only save one value, but when you switch type we revert back to the default for that type.

Yeap, I think that makes sense and is better than the current situation.

ByteHamster commented on playback/service/src/main/java/de/danoeh/antennapod/playback/service/internal/ClockSleepTimer.java:
```diff
@@ -134,4 +136,19 @@ public void updateRemainingTime(long waitingTimeOrEpisodes) {
+
+    @Override
+    public boolean isEndingThisEpisode(long episodeRemainingMillis) {
+        return episodeRemainingMillis >= getTimeLeft().getMillisValue();
```
What? :D

ByteHamster commented:
I requested it from copilot :) I didn't request it from you intentionally

keunes commented:
Thanks a lot @eblis!

eblis commented:
Thank you as well for taking the time to look into this and finally approving an episode sleep timer 😄 

I'm not done yet though, I still want to create the pull request for the UI changes for sleep timers, so they're more visible than they are right now.
I know it's still in discussion and the final UI hasn't been decided yet, but I'll start the pull request and we can go from there.
